core/src/avm2/verify.rs

```diff
@@ -0,0 +1,1193 @@
+    i: i32,
+    offset: i32,
+    one_off: bool,
+    idx_to_byte_offset: &[i32],
+    byte_offset_to_idx: &HashMap<i32, i32>,
+) -> Result<i32, Error<'gc>> {
+    let byte_offset = if one_off {
+        idx_to_byte_offset.get((i + 1) as usize).unwrap()
+    } else {
+        idx_to_byte_offset.get(i as usize).unwrap()
+    };
+    let new_byte_offset = byte_offset + offset;
+    let new_idx = byte_offset_to_idx
+        .get(&new_byte_offset)
+        .copied()
+        .ok_or_else(|| {
+            Error::AvmError(verify_error(
+                activation,
+                "Error #1021: At least one branch target was not on a valid instruction in the method.",
+                1021,
+            ).expect("Error should construct"))
+        })?;
+
+    Ok(new_idx - 1)
+}
+
+fn verify_code_starting_from<'gc>(
+    activation: &mut Activation<'_, 'gc>,
+    method: &BytecodeMethod<'gc>,
+    idx_to_byte_offset: &[i32],
+    byte_offset_to_idx: &HashMap<i32, i32>,
+    verified_blocks: &mut Vec<i32>,
+    ops: &[AbcOp],
+    start_idx: i32,
+) -> Result<(), Error<'gc>> {
+    let body = method
+        .body()
+        .expect("Cannot verify non-native method without body!");
+    let max_locals = body.num_locals;
+
+    let mut worklist = Vec::new();
+    worklist.push(start_idx);
+
+    while !worklist.is_empty() {
+        let mut save_current_work = false;
+        let mut i = *worklist.last().unwrap();
+        loop {
+            if (i as usize) >= ops.len() {
+                return Err(Error::AvmError(verify_error(
+                    activation,
+                    "Error #1020: Code cannot fall off the end of a method.",
+                    1020,
+                )?));
+            }
+
+            let op = &ops[i as usize];
+
+            // Special control flow ops
+            match op {
+                AbcOp::IfEq { offset }
+                | AbcOp::IfFalse { offset }
+                | AbcOp::IfGe { offset }
+                | AbcOp::IfGt { offset }
+                | AbcOp::IfLe { offset }
+                | AbcOp::IfLt { offset }
+                | AbcOp::IfNe { offset }
+                | AbcOp::IfNge { offset }
+                | AbcOp::IfNgt { offset }
+                | AbcOp::IfNle { offset }
+                | AbcOp::IfNlt { offset }
+                | AbcOp::IfStrictEq { offset }
+                | AbcOp::IfStrictNe { offset }
+                | AbcOp::IfTrue { offset }
+                | AbcOp::Jump { offset } => {
+                    let op_idx = adjust_jump_offset(
+                        activation,
+                        i,
+                        *offset,
+                        true,
+                        idx_to_byte_offset,
+                        byte_offset_to_idx,
+                    )?;
+
+                    if op_idx != i {
+                        // Replace last entry on worklist with current i (update i)
+                        worklist.pop();
+                        worklist.push(i + 1);
```

Comment:
I don't really understand what this (and `save_current_work` are for). Is there anything that makes the naive impl, like
```rs
    while let Some(mut i) = worklist.pop() {
        // no other popping inside
```
not work?

Reply:
This is mostly because on the conditional jumps, we need to push `i` back on the worklist.